hls-plugin-api/src/Ide/Plugin/Properties.hs

```diff
@@ -0,0 +1,341 @@
+  SBoolean :: SPropertyKey (PropertyKey s 'TBoolean)
+  SObject :: SPropertyKey (PropertyKey s 'TObject)
+  SArray :: SPropertyKey (PropertyKey s 'TArray)
+  SEnum :: SPropertyKey (PropertyKey s 'TEnum)
+
+data SomePropertyKeyWithMetaData
+  = forall k s t.
+    (k ~ PropertyKey s t) =>
+    SomePropertyKeyWithMetaData (SPropertyKey k, MetaData t)
+
+-- | Describes dedicated configuration of a plugin.
+--
+-- It was designed to be compatible with vscode's poor settings UI
+newtype Properties (r :: [*]) = Properties (Map.Map String SomePropertyKeyWithMetaData)
+
+-- ---------------------------------------------------------------------
+
+type family IsTEnum (t :: PropertyType) :: Constraint where
+  IsTEnum 'TEnum = ()
+  IsTEnum x = TypeError ('Text "Expected ‘" ':<>: 'ShowType 'TEnum ':<>: 'Text "’, but got ‘" ':<>: 'ShowType x ':<>: 'Text "’")
+
+type family IsNotTEnum (t :: PropertyType) :: Constraint where
+  IsNotTEnum 'TEnum = TypeError ('Text "Unexpected " ':<>: 'ShowType 'TEnum)
+  IsNotTEnum x = ()
+
+type family FindKey (s :: Symbol) r where
+  FindKey s (PropertyKey s t ': _) = t
+  FindKey s (_ ': xs) = FindKey s xs
+
+type family Elem (s :: Symbol) r :: Constraint where
+  Elem s (PropertyKey s _ ': _) = ()
+  Elem s (_ ': xs) = Elem s xs
+  Elem s '[] = TypeError ('Text "The key ‘" ':<>: 'Text s ':<>: 'Text "’ is undefined")
+
+type family NotElem (s :: Symbol) r :: Constraint where
+  NotElem s (PropertyKey s _ ': _) = TypeError ('Text "The key ‘" ':<>: 'Text s ':<>: 'Text "’ is already defined")
+  NotElem s (_ ': xs) = NotElem s xs
+  NotElem s '[] = ()
+
+type HasProperty s k t r = (k ~ PropertyKey s t, Elem s r, FindKey s r ~ t, KnownSymbol s)
+
+-- ---------------------------------------------------------------------
+
+emptyProperties :: Properties '[]
+emptyProperties = Properties Map.empty
+
+insert ::
+  forall s k r t.
+  (k ~ PropertyKey s t, NotElem s r, KnownSymbol s) =>
+  SPropertyKey k ->
+  MetaData t ->
+  Properties r ->
+  Properties (k ': r)
+insert key metadata (Properties old) = Properties (Map.insert (symbolVal (Proxy @s)) (SomePropertyKeyWithMetaData (key, metadata)) old)
+
+find ::
+  forall s k r t.
+  (HasProperty s k t r) =>
+  Properties r ->
+  (SPropertyKey k, MetaData t)
+find (Properties p) = case p Map.! symbolVal (Proxy @s) of
+  (SomePropertyKeyWithMetaData x) ->
+    -- Note [Constraints]
+    -- It's safe to use unsafeCoerce here:
+    --   Since each property name is unique that the redefinition will be prevented by predication on the type level list,
+    --   the value we get from the name-indexed map must be exactly the singleton and metadata corresponding to the type.
+    -- We drop this information at type level: some of the above type families return '() :: Constraint',
+    -- so GHC will consider them as redundant.
+    -- But we encode it using semantically identical 'Map' at term level,
+    -- which avoids inducting on the list by defining a new type class.
+    unsafeCoerce x
```

Comment:
I'm sure this is probably fine, but you don't need the unsafe coerce here. You can instead pack a proof in `SomePropertyKeyWithMetadata` that this thing is where you say it is in the row.

Reply:
Might not be worth doing though, unless you want to flex your type-level :muscle: s